### ISSUE TITLE
Incompatible with PHP 5.3.*

### ISSUE THREAD

**Issue by pableu** [phpseclib/phpseclib#1043]
72d1bdf60be855d breaks PHP 5.3 compatibility for me. 

    Parse error: syntax error, unexpected '[' in /www/vendor/phpseclib/phpseclib/phpseclib/Math/BigInteger.php on line 3771

I suggest either making it 5.3 compatible again or updating composer.json to reflect the increased version requirements.

**Comment by terrafrost** [comment]
The master branch should not be used in a production environment. The changes I have locally (that are still under development) will up the version requirement to 7.0.

72d1bdf is not a change that is present in 1.0 or 2.0. Just master.

**Comment by Sxderp** [comment]
Upping the requirement to 7.0 might be a little harsh. What's the argument for not sticking with at least 5.6.x?

**Comment by terrafrost** [comment]
Well this is for the master branch. Eventually I'll create a 3.0.0 branch from the master branch but I'm not at that point yet. And when 3.0.0 is released PHP 7 might not be so new.

But the reason I have it upped to PHP 7 in the master branch is because I'm using "Constant arrays using define()", which is discussed at http://php.net/manual/en/migration70.new-features.php. I'm using `define()` instead of `const` because you can't do array unions in `const`'s whereas you can in `define()`. Maybe I'll change how I'm doing things as I rewrite additional parts of the code. The minimum required version could go up or down depending on how things go during various refactors / rewrites.

**Comment by pableu** [comment]
We used master because of a bug in 2.0.1 that prevented us from using it. The bug is fixed in 2.0.2, so we switched our code back to using master. The issue can be closed for me. 

But again, wouldn't it make sense to have a composer.json in master that reflects the current version requirement of master?

**Comment by terrafrost** [comment]
> @terrafrost Please use const instead of define as the latter is in global space.

I actually am namespacing the define's. eg.

```php
define(__NAMESPACE__ . '\PrivateKeyInfo', [
    'type'     => ASN1::TYPE_SEQUENCE,
    'children' => [
        'version' => Version,
        'privateKeyAlgorithm'=> AlgorithmIdentifier,
        'privateKey' => PrivateKey,
        'attributes' => [
            'constant' => 0,
            'optional' => true,
            'implicit' => true
        ] + Attributes
    ]
]);
```

But you see that `'privateKey' => PrivateKey,` bit? PrivateKey is another constant. I can't reference constants in constants with `const` but I can with `define`.

The thing is...  with the work that I've done ASN1.php will be used in a -lot- more places. And ASN1 involves massive arrays. But portions of the arrays can be re-used. Like AlgorithmIdentifier...  that's used all over the place.

In 1.0 and 2.0 X509.php has a ton of little arrays defined that incorporate other arrays into them. But that's all done in the constructor, which I'm not a big fan of. I don't want to have to do  `$pkcs8 = new PKCS8(); ASN1::asn1map($decoded[0], $pkcs8->PrivateKeyInfo)` (`asn1map` isn't currently static in my changes but I figure I'll tackle that in a later change). Right now what you'd have to do is `ASN1::asn1map($decoded[0], \phpseclib\Crypt\RSA\Common\PrivateKeyInfo)` and `\phpseclib\Crypt\RSA\Common\PrivateKeyInfo` is an array.

Camel case constants using `define()` is also a bit weird. The convention is to normally have them be all upper case. But it's consistent with the ASN1 specs in the various RFC's I'm consulting.
